File: src/main/java/com/scottmo/ui/components/FilePicker.java
package com.scottmo.ui.components;

import com.scottmo.Application;
import com.scottmo.Config;
import net.miginfocom.swing.MigLayout;

import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JFileChooser;
import javax.swing.JLabel;
import javax.swing.JList;
import javax.swing.JScrollPane;
import javax.swing.JTextField;
import javax.swing.ListSelectionModel;
import java.awt.Dialog;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.io.File;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Vector;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public final class FilePicker {
    public static final int FILES = 0;
    public static final int DIRECTORIES = 1;
    public static final int FILES_AND_DIRECTORIES = 2;
    public static final int FILE_SEARCH = 3;

    private static String selectedFilePath = "";

    public void show(Consumer<String> onSelected) {
        show(FILES_AND_DIRECTORIES, Config.get().dataDir(), onSelected);
    }

    public static void show(int mode, String dirPath, Consumer<String> onSelected) {

        if (mode != FILE_SEARCH) {
            showSystemPicker(mode, dirPath, onSelected);
            return;
        }

        // components
        var modalDialog = new JDialog(Application.get(), "Select File", Dialog.ModalityType.DOCUMENT_MODAL);
        var searchInput = new JTextField();
        var fileList = new JList<String>();
        var selectBtn = new JButton("Select");
        var cancelBtn = new JButton("Cancel");

        // view
        searchInput.setColumns(20);

        fileList.setFixedCellHeight(16);
        fileList.setFixedCellWidth(400);
        fileList.setVisibleRowCount(10);
        fileList.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);

        int appHeight = Application.get().getHeight();
        int appWidth = Application.get().getWidth();
        modalDialog.setBounds(appWidth / 4, appHeight / 4, appWidth / 2, appHeight / 2);

        var modalDialogContent = modalDialog.getContentPane();
        modalDialogContent.setLayout(new MigLayout("wrap 5"));
        modalDialogContent.add(new JLabel("Search"));
        modalDialogContent.add(searchInput, "span, align left");
        modalDialogContent.add(new JScrollPane(fileList), "span");
        modalDialogContent.add(selectBtn);
        modalDialogContent.add(cancelBtn);

        // controls
        List<String> fileNames = listFiles(dirPath);
        fileList.setListData(new Vector<>(fileNames));

        fileList.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseReleased(MouseEvent me) {
                selectBtn.setEnabled(true);
                selectedFilePath = fileList.getSelectedValue();
            }
        });

        searchInput.addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                if (e.getKeyCode() != KeyEvent.VK_ENTER) return;

                String searchText = searchInput.getText().toLowerCase();
                fileList.setListData(new Vector<>(fileNames.stream()
                        .filter(name -> name.toLowerCase().contains(searchText))
                        .collect(Collectors.toList())));
            }
        });

        selectBtn.setEnabled(false);
        selectBtn.addActionListener(e -> {
            onSelected.accept(Path.of(dirPath, selectedFilePath).toString());
            modalDialog.dispose();
        });

        cancelBtn.addActionListener(e -> {
            modalDialog.dispose();
        });

        modalDialog.setVisible(true);
    }

    private static List<String> listFiles(String dataPath) {
        File[] files = new File(Path.of(dataPath).toString()).listFiles();
        if (files == null) return Collections.emptyList();

        return Arrays.stream(files)
                .map(File::getName)
                .sorted()
                .collect(Collectors.toList());
    }

    private static void showSystemPicker(int mode, String defaultPath, Consumer<String> onSelected) {
        var startPath = Path.of(defaultPath).toFile();
        if (!startPath.isDirectory()) {
            startPath = startPath.getParentFile();
        }
        var fc = new JFileChooser();
        fc.setCurrentDirectory(startPath);
        fc.setFileSelectionMode(mode);
        var returnVal = fc.showSaveDialog(Application.get());
        if (returnVal == JFileChooser.APPROVE_OPTION) {
            var selectedPath = fc.getSelectedFile().toPath().toAbsolutePath().toString();
            onSelected.accept(selectedPath);
        }
    }
}

File: src/main/kotlin/com/scottscmo/ui/FilePicker.kt
package com.scottscmo.ui

import com.scottscmo.Application
import com.scottscmo.Config
import net.miginfocom.swing.MigLayout
import java.awt.Dialog
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.io.File
import java.io.IOException
import java.nio.file.Path
import java.util.*
import javax.swing.*

object FilePicker {
    fun show(mode: String = "fileAndDirectory", dirPath: String = Config.get().dataDir(),
             onSelected: (filePath: String) -> Unit) {

        if (mode != "filelist") {
            showSystemPicker(mode, dirPath, onSelected)
            return
        }

        val appHeight = Application.get().height
        val appWidth = Application.get().width

        val searchInput = JTextField()
        val fileList = JList(emptyArray<String>())
        val selectBtn = JButton("Select")
        val cancelBtn = JButton("Cancel")
        var selectedFilePath = ""

        val modalDialog = JDialog(Application.get(), "Select File", Dialog.ModalityType.DOCUMENT_MODAL)
        modalDialog.contentPane.apply {
            layout = MigLayout("wrap 5")
            add(JLabel("Search"))
            add(searchInput.apply {
                columns = 20
            }, "span, align left")
            add(JScrollPane(fileList.apply {
                fixedCellHeight = 16
                fixedCellWidth = 400
                visibleRowCount = 10
                selectionMode = ListSelectionModel.SINGLE_SELECTION
            }), "span")
            add(selectBtn)
            add(cancelBtn)
        }

        val fileNames = listFiles(dirPath)
        fileList.setListData(fileNames.toTypedArray())

        fileList.addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(me: MouseEvent) {
                selectBtn.isEnabled = true
                selectedFilePath = fileList.selectedValue
            }
        })

        searchInput.addKeyListener(object : KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                if (e.keyCode != KeyEvent.VK_ENTER) return

                val searchText = searchInput.text.lowercase(Locale.getDefault())
                fileList.setListData(fileNames
                    .filter { name -> name.lowercase(Locale.getDefault()).contains(searchText) }
                    .toTypedArray())
            }
        })

        selectBtn.isEnabled = false
        selectBtn.addActionListener {
            onSelected(Path.of(dirPath, selectedFilePath).toString())
            modalDialog.dispose()
        }

        cancelBtn.addActionListener {
            modalDialog.dispose()
        }

        modalDialog.setBounds(appWidth / 4, appHeight / 4, appWidth / 2, appHeight / 2)
        modalDialog.isVisible = true
    }

    private fun listFiles(dataPath: String): List<String> {
        return try {
            File(Path.of(dataPath).toString()).listFiles()
                ?.map { f -> f.name }
                ?.sorted()
                ?: emptyList()
        } catch (e: IOException) {
            System.err.println(e.message)
            listOf("Unable to load files!")
        }
    }

    private fun showSystemPicker(mode: String = "fileAndDirectory", defaultPath: String = Config.get().dataDir(),
                                 onSelected: (filePath: String) -> Unit) {
        var startPath = Path.of(defaultPath).toFile()
        if (!startPath.isDirectory) {
            startPath = startPath.parentFile
        }
        val fc = JFileChooser().apply {
            currentDirectory = startPath
            fileSelectionMode = when (mode) {
                "file" -> JFileChooser.FILES_ONLY
                "directory" -> JFileChooser.DIRECTORIES_ONLY
                else -> JFileChooser.FILES_AND_DIRECTORIES
            }
        }
        val returnVal = fc.showSaveDialog(Application.get())
        if (returnVal == JFileChooser.APPROVE_OPTION) {
            val selectedPath = fc.selectedFile.toPath().toAbsolutePath().toString()
            onSelected(selectedPath)
        }
    }
}
